tradeoffs.Rmd

```diff
@@ -0,0 +1,172 @@
+Mocks, especially, can help you write isolated unit tests.
+
+On the other hand, [*integration tests*](https://en.wikipedia.org/wiki/Integration_testing) should test how your code interacts with other software or services.
+["Narrow" integration tests](https://martinfowler.com/bliki/IntegrationTest.html) are run against test doubles, typically fakes.
+This may be most helpful to test more complex, perhaps stateful, interactions between your code and the API, which relies on internal logic of the API rather than a one-off response.
+For example, you may want to test whether if you submit a string to a translation service, you get back a result in the appropriate language.
+Such test duplication of external business logic may be hard to get right and maintain, and should rarely be necessary.
+
+["Broad" integration tests](https://martinfowler.com/bliki/IntegrationTest.html) are run against live services, and thus require no test doubles.
+
+Neither unit nor integration tests need to cover an API specification comprehensively.
+Testing the API is not your concern, but merely the narrow interface between the API and your code.
+As a rule of thumb, your code should handle, and your tests should cover the failure *modes* of the API, including HTTP response codes, API response codes or incorrect response formats.
+Perhaps you can additionally test some API requests against some known outputs, but this will always remain a sample.
+There's always a chance that some deeply nested JSON response will break your code.
+You may not be able to cover all such errors from an errand or changed API, and that's okay.
+
+## Example
+
+To illustrate these tradeoffs, consider the [biblids](http://subugoe.github.io/biblids/) package, which wraps the [DOI resolution proxy server REST API](https://www.doi.org/factsheets/DOIProxy.html#rest-api) at doi.org.
+Digital object identifiers (DOIs) are used to persistently identify scientific publications such as `10.1038/nature14236` for an article in the journal Nature.
+The doi.org system lets institutions register such (persistent) DOIs, and update (non-persistent) URLs where full texsts can currently be found.
+It might be helpful to query this API from R.
+
+Happily, the doi.org API is unauthenticated, highly performant and gives simple responses.
+For any given `GET` request, there are only these outcomes:
+
+1. A DOI cannot be found DOI.org (HTTP status code 402).
+2. A DOI can be found (HTTP status code 200).
+    a. The queried value can be found (API response code 1).
+    b. The queried value cannot be found (API response code 200).
+       For example, a DOI is registered but has no URL to resolve to.
+       Rare but theoretically possible.
+3. Something unspecified went wrong on the server end (HTTP status code 500).
+
+Omitting some bells and whistles from biblids, we can minimally wrap the API thus:
+
+```{r minimal-wrapper}
+# unexported helper
+# this is not tested, b/c the implementation could change
```

Comment:
why is it not tested?

Reply:
I mean it's tested _indirectly_?

Reply:
Hopefully explained better what I meant: internal functions aren't supposed to have tests.

Reply:
I debated adding roxygen tags to the example, but that just bloated it too much and is a separate concern anyway.